spec/support/macros.rb

```diff
@@ -0,0 +1,26 @@
+# 这里表示rspec的宏指令，可以在其它rspec方法中调用这里定义的方法。
+
+def set_todo_current_user
+  john = Fabricate(:user)
+  session[:user_id] = john.id
+end
+
+def current_user
+  User.find(session[:user_id])
+end
+
+def clear_current_user
+  session[:user_id] = nil
+end
+
+def set_current_user(user=nil)
+  session[:user_id] = ( user || Fabricate(:user) ).id
+end
+
+def sign_in(a_user=nil)
+  user = a_user || Fabricate(:user)
```

Comment:
You can use conditional assignment here.

```ruby
def sign_in(user = nil)
  user ||= Fabricate(:user)
  # ...
```

Check out memoization as well (even this is not entirely memoization its very close).